vita3k/host/include/host/pup_types.h

```diff
@@ -0,0 +1,610 @@
+    PERMISSIONS_4 = 0x4,
+    TYPE_5 = 0x5,
+    TYPE_6 = 0x6,
+    TYPE_7 = 0x7,
+    SYSCON_8 = 0x8,
+    BOOT = 0x9,
+    VS0 = 0xA,
+    CPFW = 0xB,
+    MOTION_C = 0xC,
+    BBMC_D = 0xD,
+    TYPE_E = 0xE,
+    MOTION_F = 0xF,
+    TOUCH_10 = 0x10,
+    TOUCH_11 = 0x11,
+    SYSCON_12 = 0x12,
+    SYSCON_13 = 0x13,
+    SYSCON_14 = 0x14,
+    TYPE_15 = 0x15,
+    VS0_TAR_PATCH = 0x16,
+    SA0 = 0x17,
+    PD0 = 0x18,
+    SYSCON_19 = 0x19,
+    TYPE_1A = 0x1A,
+    PSPEMU_LIST = 0x1B
+};
+
+enum class ControlType {
+    CONTROL_FLAGS = 1,
+    DIGEST_SHA1 = 2,
+    NPDRM_PS3 = 3,
+    DIGEST_SHA256 = 4,
+    NPDRM_VITA = 5,
+    UNK_SIG1 = 6,
+    UNK_HASH1 = 7
+};
+
+enum class SecureBool {
+    UNUSED = 0,
+    NO = 1,
+    YES = 2
+};
+
+enum class EncryptionType {
+    NONE = 1,
+    AES128CTR = 3
+};
+
+enum class HashType {
+    NONE = 1,
+    HMACSHA1 = 2,
+    HMACSHA256 = 6
+};
+
+enum class CompressionType {
+    NONE = 1,
+    DEFLATE = 2
+};
+
+class KeyStore {
+private:
+    std::unordered_map<
+    int, std::unordered_map<
+    int, std::unordered_map<
+    int, std::vector<KeyEntry>>>> _store;
+
+public:
+    void register_keys(KeyType keytype, SceType scetype, int keyrev, std::string key, std::string iv, uint64_t minver = 0, uint64_t maxver = 0xffffffffffffffff, SelfType selftype = SelfType::NONE) {
+        _store[static_cast<int>(keytype)][static_cast<int>(scetype)][static_cast<int>(selftype)].push_back({ minver, maxver, keyrev, key, iv });
+    }
+
+    KeyEntry get(KeyType keytype, SceType scetype, uint64_t sysver = -1, int keyrev = -1, SelfType selftype = SelfType::NONE) {
+        if (_store.find(static_cast<int>(keytype)) == _store.end()) {
+            LOG_ERROR("Cannot find any keys for this key type");
+        }
+        if (_store.at(static_cast<int>(keytype)).find(static_cast<int>(scetype)) == _store.at(static_cast<int>(keytype)).end()) {
+            LOG_ERROR("Cannot find any keys for this SCE type");
+        }
+        if (_store.at(static_cast<int>(keytype)).at(static_cast<int>(scetype)).find(static_cast<int>(selftype)) == _store.at(static_cast<int>(keytype)).at(static_cast<int>(scetype)).end()) {
+            LOG_ERROR("Cannot find any keys for this SELF type");
+        }
+
+        for (auto item : _store[static_cast<int>(keytype)][static_cast<int>(scetype)][static_cast<int>(selftype)]) {
+            if (sysver < 0 || (sysver >= item.minver && sysver <= item.maxver) && (keyrev < 0 || keyrev == item.keyrev)) {
+                return item;
+            }
+        }
+        LOG_ERROR("Cannot find key/iv for this SCE file");
+    }
+};
+
+class SceHeader {
+private:
+    uint32_t magic;
+    uint32_t version;
+    uint8_t _platform;
+    uint16_t _sce_type;
+    uint64_t data_length;
+
+public:
+    static const int Size = 32;
+    SceType sce_type;
+    SelfPlatform platform;
+    uint64_t header_length;
+    uint32_t metadata_offset;
+    uint8_t key_revision;
+
+    SceHeader(char data[]) {
```

Comment:
Nice job making this function, I think this makes code very clear.

```suggestion
    explicit SceHeader(const char *data) {
```

This makes:
- Explicit: The object not construct automatically by returning or doing something like { data }; Use this when possible.
- const char *: Don't know if char data[] involves a copy or not, but this is more safe and clean.